tempesta_fw/http_parser.c

```diff
@@ -4023,17 +4025,70 @@ Req_Method_1CharStep: __attribute__((cold))
 }
 STACK_FRAME_NON_STANDARD(tfw_http_parse_req);
 
+int
+tfw_h2_parse_hdr(const char *data, unsigned long len, TfwHttpReq *req)
+{
+	/*
+	 * TODO #309: implementation of separate parsing procedure for message
+	 * headers in HTTP/2 context.
+	 */
+	return T_OK;
+}
+
+static int
+tfw_h2_parse_body(const char *data, unsigned long len, TfwHttpReq *req,
+		  unsigned int *parsed)
+{
+	/*
+	 * TODO #309: implementation of message body parsing (i.e. payload of
+	 * HTTP/2 DATA frame).
+	 */
+	return T_OK;
+}
+
+static inline int
+tfw_h2_parse_fin(TfwHttpReq *req)
+{
+	TfwStr *crlf = &req->crlf;
+	TfwMsgParseIter *it = &req->pit;
+
+	WARN_ON_ONCE(!TFW_STR_EMPTY(crlf));
+	BUFFER_GET(2, it);
+	if (!it->pos)
+		return -ENOMEM;
+
+	crlf->data = it->pos;
+	crlf->len = 2;
+	crlf->flags |= TFW_STR_COMPLETE;
+
+	memcpy_fast(it->pos, "\r\n", 2);
```

Comment:
Done.

Reply:
By the way, GCC is smart enough to make that optimization if one uses `memcpy()`:
```
#include <string.h>

void func1(char *pos) {
    memcpy(pos, "\r\n", 2);
}
void func2(char *pos) {
    *(short *)pos = *(short *)"\r\n";
}
```
compiles to:
```
func1:
        movl    $2573, %eax
        movw    %ax, (%rdi)
        ret
func2:
        movl    $2573, %eax
        movw    %ax, (%rdi)
        ret
```

Reply:
I wasn't (and I'm still) sure about the code generation for the kernel. I checked the kernel compiler options and didn't see any prohibition of `memcpy()` optimizations. But since the kernel provides it's own version of `memcpy()`, I'm careful about optimization assumptions.